src/Rules/Classes/ClassExtendsInternalClassRule.php

```diff
@@ -0,0 +1,91 @@
+<?php declare(strict_types=1);
+
+namespace mglaman\PHPStanDrupal\Rules\Classes;
+
+use PhpParser\Node;
+use PhpParser\Node\Stmt\Class_;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\ReflectionProvider;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+
+class ClassExtendsInternalClassRule implements Rule
+{
+    /**
+     * @var ReflectionProvider
+     */
+    private $reflectionProvider;
+
+    public function __construct(ReflectionProvider $reflectionProvider)
+    {
+        $this->reflectionProvider = $reflectionProvider;
+    }
+
+    public function getNodeType(): string
+    {
+        return Class_::class;
+    }
+
+    public function processNode(Node $node, Scope $scope): array
+    {
+        /** @var Class_ $node */
+        if (!isset($node->extends)) {
+            return [];
+        }
+
+        $extendedClassName = $node->extends->toString();
+        if (!$this->reflectionProvider->hasClass($extendedClassName)) {
+            return [];
+        }
+
+        $extendedClassReflection = $this->reflectionProvider->getClass($extendedClassName);
+        if (!$extendedClassReflection->isInternal()) {
+            return [];
+        }
+
+        if (!isset($node->namespacedName)) {
+            return $this->buildError(null, $extendedClassName);
+        }
+
+        $currentClassName = $node->namespacedName->toString();
+
+        if (!$this->isInDrupalNamespace($node)) {
+            return $this->buildError($currentClassName, $extendedClassName);
+        }
+
+        if ($this->isClassAndExtendedClassInSameDrupalProjectNamespace($node)) {
+            return [];
+        }
+
+        return $this->buildError($currentClassName, $extendedClassName);
+    }
+
+    private function buildError(?string $currentClassName, string $extendedClassName): array
+    {
+        return [
+            RuleErrorBuilder::message(\sprintf(
+                '%s extends @internal class %s.',
+                $currentClassName !== null ? \sprintf('Class %s', $currentClassName) : 'Anonymous class',
+                $extendedClassName
+            ))->build()
+        ];
+    }
+
+    private function isInDrupalNamespace(Class_ $class): bool
+    {
+        return 'Drupal' === (string) $class->namespacedName->slice(0, 1);
+    }
+
+    private function isClassAndExtendedClassInSameDrupalProjectNamespace(Class_ $class): bool
```

Comment:
I think we can rename this to `isSharedNamespace`. We consider it to be "shared" if the first two levels match:

* \Drupal\Core
* \Drupal\Component
* \Drupal\{extension}